java-checks/src/main/java/org/sonar/java/checks/spring/ControllerWithSessionAttributesCheck.java

```diff
@@ -0,0 +1,105 @@
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3753")
+public class ControllerWithSessionAttributesCheck extends AbstractRequestMappingVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    ClassTree classTree = (ClassTree) tree;
+    SymbolMetadata classMetadata = classTree.symbol().metadata();
+    Optional<AnnotationTree> sessionAttributesAnnotation = classTree.modifiers().annotations()
+      .stream()
+      .filter(a -> a.annotationType().symbolType().fullyQualifiedName().equals("org.springframework.web.bind.annotation.SessionAttributes"))
+      .findFirst();
+    if (classMetadata.isAnnotatedWith("org.springframework.stereotype.Controller")
+        && sessionAttributesAnnotation.isPresent()
+        && classTree.members().stream().noneMatch(ControllerWithSessionAttributesCheck::methodCompletesSessionStatus)) {
+      reportIssue(sessionAttributesAnnotation.get().annotationType(),
+        "Add a call to \"setComplete()\" on the SessionStatus object in a \"@RequestMapping\" method that handles \"POST\".");
+    }
+  }
+
+  private static boolean methodCompletesSessionStatus(Tree tree) {
+    if (!tree.is(Tree.Kind.METHOD)) {
+      return false;
+    }
+
+    MethodTree methodTree = (MethodTree) tree;
+    List<AnnotationTree> annotationTrees = methodTree.modifiers().annotations();
+    if (annotationTrees.stream().anyMatch(ControllerWithSessionAttributesCheck::isPostRequest)
+        && methodTree.parameters().stream().anyMatch(p -> p.type().symbolType().fullyQualifiedName().equals("org.springframework.web.bind.support.SessionStatus"))) {
+      MethodInvocationVisitor methodInvocationVisitor = new MethodInvocationVisitor();
+      methodTree.block().accept(methodInvocationVisitor);
+      return methodInvocationVisitor.found;
+    }
+    return false;
+  }
+
+  private static boolean isPostRequest(AnnotationTree annotation) {
+    if (annotation.symbolType().is("org.springframework.web.bind.annotation.RequestMapping")) {
+      List<ExpressionTree> methodValues = annotation.arguments().stream()
+        .filter(argument -> "method".equals(attributeName(argument)))
+        .flatMap(AbstractRequestMappingVisitor::extractValues)
+        .collect(Collectors.toList());
+      if (methodValues.size() == 1) {
+        return getRequestMethodEnumEntry(methodValues.get(0)).equals("POST");
+      }
+      return false;
+    } else {
+      return annotation.symbolType().is("org.springframework.web.bind.annotation.PostMapping");
+    }
+  }
+
+  private static class MethodInvocationVisitor extends BaseTreeVisitor {
+    boolean found;
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (tree.symbol().toString().equals("SessionStatus#setComplete()")) {
```

Comment:
It's a fragile way to test "setComplete()", you should use:
SET_COMPLETE_MATCHER = MethodMatcher.create().typeDefinition("org.springframework.web.bind.support.SessionStatus").name("setComplete").withoutParameter());
and then:
SET_COMPLETE_MATCHER.matches(tree)